crates/cubecl-wgpu/src/compiler/base.rs

```diff
@@ -1,31 +1,113 @@
+        }
+    }
+}
+
+impl Display for DynRepresentation {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DynRepresentation::Wgsl(compute_shader) => compute_shader.fmt(f),
+            #[cfg(feature = "spirv")]
+            DynRepresentation::SpirV(spirv_kernel) => spirv_kernel.fmt(f),
+        }
+    }
+}
+
+impl CompilerRepresentation for DynRepresentation {
+    fn shared_memory_size(&self) -> usize {
+        0
+    }
+}
+
+impl Compiler for DynCompiler {
+    type Representation = DynRepresentation;
+
+    type CompilationOptions = WgpuCompilationOptions;
 
-pub trait WgpuCompiler: Compiler {
     fn compile(
-        server: &mut WgpuServer<Self>,
-        kernel: <WgpuServer<Self> as ComputeServer>::Kernel,
+        &mut self,
+        kernel: KernelDefinition,
+        compilation_options: &Self::CompilationOptions,
+        mode: ExecutionMode,
+    ) -> Self::Representation {
+        match self {
+            DynCompiler::Wgsl(wgsl_compiler) => {
+                Compiler::compile(wgsl_compiler, kernel, compilation_options, mode).into()
+            }
+            #[cfg(feature = "spirv")]
+            DynCompiler::SpirV(spirv_compiler) => {
+                Compiler::compile(spirv_compiler, kernel, compilation_options, mode).into()
+            }
+        }
+    }
+
+    fn elem_size(&self, elem: cubecl_core::ir::Elem) -> usize {
+        match self {
+            DynCompiler::Wgsl(wgsl_compiler) => wgsl_compiler.elem_size(elem),
+            #[cfg(feature = "spirv")]
+            DynCompiler::SpirV(spirv_compiler) => spirv_compiler.elem_size(elem),
+        }
+    }
+}
+
+impl DynCompiler {
+    pub fn compile_dyn(
+        &mut self,
+        server: &mut WgpuServer,
+        kernel: <WgpuServer as ComputeServer>::Kernel,
         mode: ExecutionMode,
-    ) -> CompiledKernel<Self>;
+    ) -> CompiledKernel<Self> {
+        match self {
+            DynCompiler::Wgsl(_) => {
+                <WgslCompiler as WgpuCompiler>::compile(self, server, kernel, mode)
+            }
+            #[cfg(feature = "spirv")]
+            DynCompiler::SpirV(_) => {
+                <super::spirv::VkSpirvCompiler as WgpuCompiler>::compile(self, server, kernel, mode)
+            }
+        }
+    }
+}
 
-    fn create_pipeline(
-        server: &mut WgpuServer<Self>,
-        kernel: CompiledKernel<Self>,
+pub trait WgpuCompiler: Compiler {
+    fn compile(
+        dyn_comp: &mut DynCompiler,
+        server: &mut WgpuServer,
+        kernel: <WgpuServer as ComputeServer>::Kernel,
         mode: ExecutionMode,
-    ) -> Arc<ComputePipeline>;
-
-    #[allow(async_fn_in_trait)]
-    async fn request_device(adapter: &Adapter) -> (Device, Queue);
-    fn register_features(
-        adapter: &Adapter,
-        device: &Device,
-        props: &mut DeviceProperties<Feature>,
-        comp_options: &mut Self::CompilationOptions,
-    );
+    ) -> CompiledKernel<DynCompiler>;
```

Comment:
Is this needed anymore? Could we just implement the function for the `DynCompiler` ?

Reply:
It was to allow for robustness checks and spirv dumping that are specifically done for SPIR-V, but that really belongs in just a function in the backend so I moved it.